libzecale/circuits/fields/fp12_2over3over2_gadgets.tcc

```diff
@@ -0,0 +1,366 @@
+
+namespace libzecale
+{
+
+// Fp12_2over3over2_variable methods
+
+template<typename Fp12T>
+Fp12_2over3over2_variable<Fp12T>::Fp12_2over3over2_variable(
+    libsnark::protoboard<FieldT> &pb, const std::string &annotation_prefix)
+    : libsnark::gadget<FieldT>(pb, annotation_prefix)
+    , _c0(pb, FMT(annotation_prefix, " c0"))
+    , _c1(pb, FMT(annotation_prefix, " c1"))
+{
+}
+
+template<typename Fp12T>
+Fp12_2over3over2_variable<Fp12T>::Fp12_2over3over2_variable(
+    libsnark::protoboard<FieldT> &pb,
+    const Fp12T &el,
+    const std::string &annotation_prefix)
+    : libsnark::gadget<FieldT>(pb, annotation_prefix)
+    , _c0(pb, el.c0, FMT(annotation_prefix, " c0"))
+    , _c1(pb, el.c1, FMT(annotation_prefix, " c1"))
+{
+}
+
+template<typename Fp12T>
+Fp12_2over3over2_variable<Fp12T>::Fp12_2over3over2_variable(
+    libsnark::protoboard<FieldT> &pb,
+    const Fp6_3over2_variable<Fp6T> &c0,
+    const Fp6_3over2_variable<Fp6T> &c1,
+    const std::string &annotation_prefix)
+    : libsnark::gadget<FieldT>(pb, annotation_prefix), _c0(c0), _c1(c1)
+{
+}
+
+template<typename Fp12T>
+Fp12T Fp12_2over3over2_variable<Fp12T>::get_element() const
+{
+    return Fp12T(_c0.get_element(), _c1.get_element());
+}
+
+template<typename Fp12T>
+void Fp12_2over3over2_variable<Fp12T>::generate_r1cs_witness(const Fp12T &el)
+{
+    _c0.generate_r1cs_witness(el.c0);
+    _c1.generate_r1cs_witness(el.c1);
+}
+
+// Fp12_2over3over2_square_gadget methods
+
+template<typename Fp12T>
+Fp12_2over3over2_square_gadget<Fp12T>::Fp12_2over3over2_square_gadget(
+    libsnark::protoboard<FieldT> &pb,
+    const Fp12_2over3over2_variable<Fp12T> &A,
+    const Fp12_2over3over2_variable<Fp12T> &result,
+    const std::string &annotation_prefix)
+    : libsnark::gadget<FieldT>(pb, annotation_prefix)
+    , _A(A)
+    , _result(result)
+    , _alpha(
+          pb,
+          _A._c0,
+          _A._c1,
+          _result._c1 * (FieldT("2").inverse()),
+          FMT(annotation_prefix, " _alpha"))
+    , _beta(
+          pb,
+          _A._c0 + _A._c1,
+          _A._c0 +
+              mul_by_non_residue(pb, _A._c1, FMT(annotation_prefix, " a1*v")),
+          _result._c0 +
+              mul_by_non_residue(
+                  pb, _alpha._result, FMT(annotation_prefix, " alpha*v")) +
+              _alpha._result,
+          FMT(annotation_prefix, " _beta"))
+{
+}
+
+template<typename Fp12T>
+const Fp12_2over3over2_variable<Fp12T>
+    &Fp12_2over3over2_square_gadget<Fp12T>::result() const
+{
+    return _result;
+}
+
+template<typename Fp12T>
+void Fp12_2over3over2_square_gadget<Fp12T>::generate_r1cs_constraints()
+{
+    _alpha.generate_r1cs_constraints();
+    _beta.generate_r1cs_constraints();
+}
+
+template<typename Fp12T>
+void Fp12_2over3over2_square_gadget<Fp12T>::generate_r1cs_witness()
+{
+    const Fp6T a0 = _A._c0.get_element();
+    const Fp6T a1 = _A._c1.get_element();
+    const Fp6T alpha = a0 * a1;
```

Comment:
FYI: `a0`, i.e. the `c0` coeff of A is denoted `Ac0` in existing gadgets in libsnark, see: https://github.com/scipr-lab/libsnark/blob/master/libsnark/gadgetlib1/gadgets/fields/fp6_gadgets.tcc#L219
That would be great to follow existing naming conventions o avoid confusion across gadget code.

Happy to handle this in a separate PR as this is not critical at this stage though, but may worth fixing going forward.

Note: Libsnark's notations are not fully consistent either since they use e.g. `Ac0` to refer to `A.c0` and also `result_c0` to refer to `result.c0`. It may be better to follow the "subscript" notation and access coefficients of extension field elements consistently with the `_c` notation.